pymbolic/primitives.py

```diff
@@ -1530,6 +1538,38 @@ def step(self):
 
     mapper_method = intern("map_slice")
 
+
+class NaN(Expression):
+    """
+    An expression node representing not-a-number as a floating point number.
+    Unlike, :data:`math.nan`, all instances of :class:`NaN` compare equal, as
+    one might reasonably expect for program representation. (If this weren't
+    so, programs containing NaNs would effectively be unhashable, because they
+    don't compare equal to themselves.)
+
+    Note that, in Python, this equality comparison is made *even* more
+    complex by `this issue <https://bugs.python.org/issue21873>`__, due
+    to which ``np.nan == np.nan`` is *False*, but ``(np.nan,) == (np.nan,)``
+    is True.
+
+    .. attribute:: data_type
+
+        The data type used for the actual realization of the constant. Defaults
+        to *None*. If given, This must be a callable to which a NaN
+        :class:`float` can be passed to obtain a NaN of the yield the desired
+        type.  It must also be suitable for use as the second argument of
+        :func:`isinstance`.
+    """
+    init_arg_names = ("data_type", )
```

Comment:
FYI: Renamed this `data_type`, since it's definitely not a numpy dtype. (Note that numpy types and dtypes are distinct.)

Reply:
Thanks!

> Note that numpy types and dtypes are distinct

Aah yep!